### PR TITLE
Template RadialReturnStressUpdate and StressUpdateBase

### PR DESCRIPTION
**Pull request by dewenyushu**
Address issue #16183

### CODE REVIEW

**Comment by dewenyushu** [comment]
@dschwen templating `RadialReturnStressUpdate` requires some minor changes in Bison, too. The inputs in the `computeStressInitialize` member function needs to be passed by reference. I have a separate MR in Bison for this. 

So I think this PR is ready to be reviewed :)

File: modules/tensor_mechanics/include/materials/RadialReturnStressUpdate.h [hunk 1]
```diff
@@ -188,4 +245,60 @@ class RadialReturnStressUpdate : public StressUpdateBase, public SingleVariableR
...
+
+template <>
+void RadialReturnStressUpdateTempl<true>::updateStateSubstep(
+    ADRankTwoTensor & strain_increment,
+    ADRankTwoTensor & inelastic_strain_increment,
+    const ADRankTwoTensor & rotation_increment,
+    ADRankTwoTensor & stress_new,
+    const RankTwoTensor & stress_old,
+    const ADRankFourTensor & elasticity_tensor,
+    const RankTwoTensor & elastic_strain_old);
```
Comment (dschwen) [inline modules/tensor_mechanics/include/materials/RadialReturnStressUpdate.h:301]: why are these required?

File: modules/tensor_mechanics/include/materials/StressUpdateBase.h [hunk 2]
```diff
@@ -11,11 +11,7 @@
 
 #include "Conversion.h"
 #include "InputParameters.h"
-#include "Material.h"
-#include "RankFourTensor.h"
-#include "RankTwoTensor.h"
-
-// Forward declaration
+#include "ADMaterial.h"
```
Comment (dschwen) [inline modules/tensor_mechanics/include/materials/StressUpdateBase.h:14]:
```suggestion
#include "Material.h"
```

File: modules/tensor_mechanics/include/materials/StressUpdateBase.h [hunk 3]
```diff
@@ -157,3 +176,9 @@ class StressUpdateBase : public Material
   /// Name used as a prefix for all material properties related to the stress update model.
   const std::string _base_name;
 };
+
+template <>
+TangentCalculationMethod StressUpdateBaseTempl<true>::getTangentCalculationMethod();
```
Comment (dschwen) [inline modules/tensor_mechanics/include/materials/StressUpdateBase.h:181]: really necessary?

**Review by dschwen — COMMENTED** [review]
Looks good to me, apart from the explicit member function instantiations. But I might be wrong there.

File: modules/tensor_mechanics/include/materials/RadialReturnStressUpdate.h (hunk 1 above)
Comment (dewenyushu) [reply modules/tensor_mechanics/include/materials/RadialReturnStressUpdate.h:301]: The declarations of the specializations are needed here to prevent calling the default templated function. Removing these, the default function that errors out will be called.

File: modules/tensor_mechanics/include/materials/RadialReturnStressUpdate.h [hunk 4]
```diff
@@ -47,35 +60,72 @@ class RadialReturnStressUpdate : public StressUpdateBase, public SingleVariableR
...
+                                  RankTwoTensor & /*inelastic_strain_increment*/,
+                                  const RankTwoTensor & /*rotation_increment*/,
+                                  RankTwoTensor & /*stress_new*/,
+                                  const RankTwoTensor & /*stress_old*/,
+                                  const RankFourTensor & /*elasticity_tensor*/,
+                                  const RankTwoTensor & /*elastic_strain_old*/,
+                                  bool /*compute_full_tangent_operator*/,
+                                  RankFourTensor & /*tangent_operator*/) override
+  {
+    mooseError("updateStateSubstep called: the version without tangent computation should be "
```
Comment (dschwen) [inline modules/tensor_mechanics/include/materials/RadialReturnStressUpdate.h:101]: move the implementation into the .C file

File: modules/tensor_mechanics/src/materials/RadialReturnStressUpdate.C [hunk 5]
```diff
@@ -86,16 +98,17 @@ RadialReturnStressUpdate::calculateNumberSubsteps(const RankTwoTensor & strain_i
...
-RadialReturnStressUpdate::updateState(RankTwoTensor & strain_increment,
-                                      RankTwoTensor & inelastic_strain_increment,
-                                      const RankTwoTensor & /*rotation_increment*/,
-                                      RankTwoTensor & stress_new,
-                                      const RankTwoTensor & /*stress_old*/,
-                                      const RankFourTensor & elasticity_tensor,
-                                      const RankTwoTensor & elastic_strain_old,
-                                      bool compute_full_tangent_operator,
-                                      RankFourTensor & tangent_operator)
+RadialReturnStressUpdateTempl<false>::updateState(RankTwoTensor & strain_increment,
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/RadialReturnStressUpdate.C:103]: and here you add the other specialization with the error message. And you never implement the generic version of the template!

Comment (dewenyushu) [reply modules/tensor_mechanics/src/materials/RadialReturnStressUpdate.C:103]: > and here you add the other specialization with the error message. And you never implement the generic version of the template!

Thanks for the comments Daniel! Here is also where I am not very sure about.

The AD and non-AD versions of `updateState`, as well as `updateStateSubstep` require different numbers of inputs, therefore what I do is just to error out when the incorrect version of the functions are called...

Do you mean to have `RankTwoTensor`->`GenericRankTwoTensor<false>`, etc, for the non-AD version? I am not sure how that would benefit as only non-AD would call this function

**Comment by dewenyushu** [comment]
@dschwen  I have made the changes and I suppose this PR is ready :)

**Review by dschwen — APPROVED** [review]
Looks good now

**Comment by dschwen** [comment]
Bison fix is ready as well?

**Comment by tophmatthews** [comment]
The documentation `ADRadialReturnStressUpdate.md` and `RadialReturnStressUpdate.md` are different, and should be one single file now, no?

**Comment by dewenyushu** [comment]
> The documentation `ADRadialReturnStressUpdate.md` and `RadialReturnStressUpdate.md` are different, and should be one single file now, no?

Thanks @tophmatthews for pointing this out. I agree the documentation should be unified as well. These two seem to be largely the same so it should be easy. I will work on a separate PR for this.
